move computeChains to hasher, make concurrency parameter part of Opts,…

ArteMisc opened the pull request:
… switch from sync.WaitGroup so completion channel.

Moved computeChains to be a method of hasher, as it seems like a logical fit.

Replaced the implicit GOMAXPROCS use for concurrency with a parameter in Opts. Passing default Opts does not change original behaviour (still uses a single goroutine). The logic behind selection of the number of goroutines used for computations is included in the comments of the parameter.

Removed the manual copying Address data from slice to slice within the goroutines, instead passing the variable by value which automatically copies the data as Address is a fixed-size type.

Switched out the sync.WaitGroup for a completion channel. Other than being a bit more idiomatic, It seems to have an ever-so-slight influence on performance, though these measurements should be taken with a few grains of salt:

```
Waitgroup implementation:

(run 1)
BenchmarkGenPublicKey-12                            1000           1205935 ns/op            5584 B/op         15 allocs/op
BenchmarkSign-12                                    2000            592959 ns/op            5701 B/op         17 allocs/op
BenchmarkPkFromSig-12                               2000            616200 ns/op            3235 B/op         14 allocs/op
BenchmarkW4GenPublicKey-12                          2000            531721 ns/op           10768 B/op         15 allocs/op
BenchmarkW4Sign-12                                  5000            301727 ns/op           11013 B/op         17 allocs/op
BenchmarkW4PkFromSig-12                             5000            256654 ns/op            5989 B/op         14 allocs/op
BenchmarkConcurrentGenPublicKey-12                  3000            434214 ns/op            8480 B/op         37 allocs/op
BenchmarkConcurrentSign-12                          5000            366515 ns/op            8595 B/op         39 allocs/op
BenchmarkConcurrentPkFromSig-12                     3000            350751 ns/op            6131 B/op         36 allocs/op
BenchmarkConcurrentW4GenPublicKey-12                5000            362475 ns/op           13664 B/op         37 allocs/op
BenchmarkConcurrentW4Sign-12                        5000            291083 ns/op           13909 B/op         39 allocs/op
BenchmarkConcurrentW4PkFromSig-12                  10000            205041 ns/op            8885 B/op         36 allocs/op


Completion channel implementation:

(run 1)
BenchmarkGenPublicKey-12                            2000           1179505 ns/op            5664 B/op         15 allocs/op
BenchmarkSign-12                                    2000            593961 ns/op            5780 B/op         17 allocs/op
BenchmarkPkFromSig-12                               2000            602155 ns/op            3315 B/op         14 allocs/op
BenchmarkW4GenPublicKey-12                          3000            527479 ns/op           10849 B/op         15 allocs/op
BenchmarkW4Sign-12                                  5000            303341 ns/op           11093 B/op         17 allocs/op
BenchmarkW4PkFromSig-12                             5000            257788 ns/op            6069 B/op         14 allocs/op
BenchmarkConcurrentGenPublicKey-12                  3000            432230 ns/op            8582 B/op         37 allocs/op
BenchmarkConcurrentSign-12                          3000            361856 ns/op            8675 B/op         39 allocs/op
BenchmarkConcurrentPkFromSig-12                     5000            318842 ns/op            6211 B/op         36 allocs/op
BenchmarkConcurrentW4GenPublicKey-12                5000            350981 ns/op           13748 B/op         37 allocs/op
BenchmarkConcurrentW4Sign-12                        5000            275830 ns/op           13989 B/op         39 allocs/op
BenchmarkConcurrentW4PkFromSig-12                  10000            193109 ns/op            8965 B/op         36 allocs/op

(run 2)
BenchmarkGenPublicKey-12                            1000           1223681 ns/op            5664 B/op         15 allocs/op
BenchmarkSign-12                                    2000            573825 ns/op            5779 B/op         17 allocs/op
BenchmarkPkFromSig-12                               2000            595281 ns/op            3315 B/op         14 allocs/op
BenchmarkW4GenPublicKey-12                          3000            536396 ns/op           10848 B/op         15 allocs/op
BenchmarkW4Sign-12                                  5000            299896 ns/op           11094 B/op         17 allocs/op
BenchmarkW4PkFromSig-12                             5000            253187 ns/op            6069 B/op         14 allocs/op
BenchmarkConcurrentGenPublicKey-12                  3000            439728 ns/op            8597 B/op         37 allocs/op
BenchmarkConcurrentSign-12                          3000            358041 ns/op            8675 B/op         39 allocs/op
BenchmarkConcurrentPkFromSig-12                     5000            329028 ns/op            6211 B/op         36 allocs/op
BenchmarkConcurrentW4GenPublicKey-12                5000            346738 ns/op           13744 B/op         37 allocs/op
BenchmarkConcurrentW4Sign-12                        5000            281884 ns/op           13989 B/op         39 allocs/op
BenchmarkConcurrentW4PkFromSig-12                  10000            192535 ns/op            8965 B/op         36 allocs/op
```

Commit:
move computeChain to hasher, make concurrency parameter part of Opts, switch from sync.WaitGroup so completion channel.

ArteMisc commented:
For what it's worth, these are my values used for the concurrent benchmark:

```
goos: linux
goarch: amd64
numCPU: 12
GOMAXPROCS: 12
```

lentus commented on params.go:
```diff
@@ -20,6 +29,25 @@ func (Opts) HashFunc() crypto.Hash {
+		return 1
+	}
+
+	if o.Concurrency >= 0 {
```
>= should be > (for clarity)

ArteMisc replied:
agreed

lentus reviewed (changes requested):
We should update the README to reflect these changes

ArteMisc committed:
Update params.go

made implementation of routines() more consistent with the documentation provided on Opts.Concurrency.
